File: sdk/src/types/passkey.ts
import { z } from 'zod';

// ============================================================================
// Passkey Credential Types
// ============================================================================

export const PasskeyCredentialSchema = z.object({
  credentialId: z.string(),
  userId: z.string(),
  publicKeyX: z.string(),
  publicKeyY: z.string(),
  smartAccountAddress: z.string().nullable().optional(),
  displayName: z.string(),
  isActive: z.boolean(),
  createdAt: z.string(),
  lastUsedAt: z.string().nullable().optional(),
});

export type PasskeyCredential = z.infer<typeof PasskeyCredentialSchema>;

// ============================================================================
// Register Passkey
// ============================================================================

export const RegisterPasskeyParamsSchema = z.object({
  userId: z.string(),
  credentialId: z.string(),
  publicKeyX: z.string().regex(/^(0x)?[0-9a-fA-F]{1,64}$/, 'Invalid P256 x coordinate'),
  publicKeyY: z.string().regex(/^(0x)?[0-9a-fA-F]{1,64}$/, 'Invalid P256 y coordinate'),
  displayName: z.string(),
});

export type RegisterPasskeyParams = z.infer<typeof RegisterPasskeyParamsSchema>;

export const RegisterPasskeyResponseSchema = z.object({
  credentialId: z.string(),
  smartAccountAddress: z.string().nullable().optional(),
  createdAt: z.string(),
});

export type RegisterPasskeyResponse = z.infer<typeof RegisterPasskeyResponseSchema>;

// ============================================================================
// Create Passkey Wallet
// ============================================================================

export const CreatePasskeyWalletParamsSchema = z.object({
  userId: z.string(),
  credentialId: z.string(),
  publicKeyX: z.string().regex(/^(0x)?[0-9a-fA-F]{1,64}$/, 'Invalid P256 x coordinate'),
  publicKeyY: z.string().regex(/^(0x)?[0-9a-fA-F]{1,64}$/, 'Invalid P256 y coordinate'),
  displayName: z.string(),
  ownerAddress: z.string().optional(),
  salt: z.string().optional(),
});

export type CreatePasskeyWalletParams = z.infer<typeof CreatePasskeyWalletParamsSchema>;

export const CreatePasskeyWalletResponseSchema = z.object({
  credentialId: z.string(),
  smartAccountAddress: z.string(),
  publicKeyX: z.string(),
  publicKeyY: z.string(),
  isDeployed: z.boolean(),
  createdAt: z.string(),
});

export type CreatePasskeyWalletResponse = z.infer<typeof CreatePasskeyWalletResponseSchema>;

// ============================================================================
// Link Smart Account
// ============================================================================

export const LinkSmartAccountParamsSchema = z.object({
  userId: z.string(),
  credentialId: z.string(),
  smartAccountAddress: z.string(),
});

export type LinkSmartAccountParams = z.infer<typeof LinkSmartAccountParamsSchema>;

// ============================================================================
// Sign Transaction
// ============================================================================

export const PasskeySignatureSchema = z.object({
  r: z.string().regex(/^(0x)?[0-9a-fA-F]{1,64}$/, 'Invalid signature r component'),
  s: z.string().regex(/^(0x)?[0-9a-fA-F]{1,64}$/, 'Invalid signature s component'),
});

export type PasskeySignature = z.infer<typeof PasskeySignatureSchema>;

export const WebAuthnAssertionSchema = z.object({
  authenticatorData: z.string(),
  clientDataJSON: z.string(),
  signature: PasskeySignatureSchema,
  credentialId: z.string(),
});

export type WebAuthnAssertion = z.infer<typeof WebAuthnAssertionSchema>;

export const SignTransactionParamsSchema = z.object({
  userId: z.string(),
  credentialId: z.string(),
  userOperation: z.object({
    sender: z.string(),
    nonce: z.string(),
    callData: z.string(),
    callGasLimit: z.string().optional(),
    verificationGasLimit: z.string().optional(),
    preVerificationGas: z.string().optional(),
    maxFeePerGas: z.string().optional(),
    maxPriorityFeePerGas: z.string().optional(),
  }),
  assertion: WebAuthnAssertionSchema,
});

export type SignTransactionParams = z.infer<typeof SignTransactionParamsSchema>;

export const SignTransactionResponseSchema = z.object({
  userOpHash: z.string(),
  sender: z.string(),
  nonce: z.string(),
  signature: z.string(),
  status: z.string(),
});

export type SignTransactionResponse = z.infer<typeof SignTransactionResponseSchema>;

// ============================================================================
// Get Counterfactual Address
// ============================================================================

export const GetCounterfactualAddressParamsSchema = z.object({
  publicKeyX: z.string(),
  publicKeyY: z.string(),
  salt: z.string().optional(),
});

export type GetCounterfactualAddressParams = z.infer<typeof GetCounterfactualAddressParamsSchema>;

export const GetCounterfactualAddressResponseSchema = z.object({
  address: z.string(),
  isDeployed: z.boolean(),
});

export type GetCounterfactualAddressResponse = z.infer<typeof GetCounterfactualAddressResponseSchema>;

File: sdk-python/src/rampos/models/passkey.py
"""Passkey wallet models for WebAuthn-based smart accounts."""

from __future__ import annotations

from pydantic import BaseModel


class PasskeyCredential(BaseModel):
    credential_id: str
    user_id: str
    public_key_x: str
    public_key_y: str
    smart_account_address: str | None = None
    display_name: str
    is_active: bool
    created_at: str
    last_used_at: str | None = None


class RegisterPasskeyParams(BaseModel):
    user_id: str
    credential_id: str
    public_key_x: str
    public_key_y: str
    display_name: str


class RegisterPasskeyResponse(BaseModel):
    credential_id: str
    smart_account_address: str | None = None
    created_at: str


class CreatePasskeyWalletParams(BaseModel):
    user_id: str
    credential_id: str
    public_key_x: str
    public_key_y: str
    display_name: str
    owner_address: str | None = None
    salt: str | None = None


class CreatePasskeyWalletResponse(BaseModel):
    credential_id: str
    smart_account_address: str
    public_key_x: str
    public_key_y: str
    is_deployed: bool
    created_at: str


class LinkSmartAccountParams(BaseModel):
    user_id: str
    credential_id: str
    smart_account_address: str


class PasskeySignature(BaseModel):
    r: str
    s: str


class WebAuthnAssertion(BaseModel):
    authenticator_data: str
    client_data_json: str
    signature: PasskeySignature
    credential_id: str


class SignTransactionUserOp(BaseModel):
    sender: str
    nonce: str
    call_data: str
    call_gas_limit: str | None = None
    verification_gas_limit: str | None = None
    pre_verification_gas: str | None = None
    max_fee_per_gas: str | None = None
    max_priority_fee_per_gas: str | None = None


class SignTransactionParams(BaseModel):
    user_id: str
    credential_id: str
    user_operation: SignTransactionUserOp
    assertion: WebAuthnAssertion


class SignTransactionResponse(BaseModel):
    user_op_hash: str
    sender: str
    nonce: str
    signature: str
    status: str


class GetCounterfactualAddressParams(BaseModel):
    public_key_x: str
    public_key_y: str
    salt: str | None = None


class GetCounterfactualAddressResponse(BaseModel):
    address: str
    is_deployed: bool
